reflectionclient/reflectionclient.go

```diff
@@ -0,0 +1,150 @@
+type ClientConfig struct {
+	Host     string
+	Insecure bool
+	Audience string
+	Timeout  time.Duration
+}
+
+// NewClient creates and returns a reflection client.
+func NewClient(ctx context.Context, config *ClientConfig) (*Client, error) {
+	var d grpc.DialOption
+	var ts oauth2.TokenSource
+
+	if !config.Insecure {
+		certPool, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, fmt.Errorf("x509.SystemCertPool failed: %w", err)
+		}
+		if certPool == nil {
+			return nil, fmt.Errorf("x509.SystemCertPool failed: nil certpool")
+		}
+		c := &tls.Config{RootCAs: certPool, MinVersion: tls.VersionTLS12}
+		tlsCreds := credentials.NewTLS(c)
+		d = grpc.WithTransportCredentials(tlsCreds)
+
+		// https://pkg.go.dev/golang.org/x/oauth2/google#FindDefaultCredentialsWithParams
+		// In a non-GCP environment, we need to place the JSON key to where
+		// GOOGLE_APPLICATION_CREDENTIALS env var is pointed to.
+		cred, err := google.FindDefaultCredentials(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate default credentials: %w", err)
+		}
+
+		aud := config.Audience
+		if aud == "" {
+			aud = config.Host
+		}
+		t, err := idtoken.NewTokenSource(ctx, aud, option.WithCredentials(cred))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create NewTokenSource: %w", err)
+		}
+		ts = t
+	} else {
+		d = grpc.WithTransportCredentials(insecure.NewCredentials())
+	}
+
+	conn, err := grpc.DialContext(ctx, config.Host, d)
+	if err != nil {
+		return nil, fmt.Errorf("grpc.DialContext failed: %w", err)
+	}
+	return &Client{
+		conn:   conn,
+		config: config,
+		ts:     ts,
+	}, nil
+}
+
+// CallMethod provides a payload to specified method, and returns the results as txt or json.
+func (c *Client) CallMethod(ctx context.Context, method, payload string, textFormat bool) (*bytes.Buffer, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, c.config.Timeout)
+	defer cancel()
+
+	var headers []string
+	// If the token source is present, generate an ID token to authenticate with
+	// the server.
+	if c.ts != nil {
+		t, err := c.ts.Token()
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate ID token: %w", err)
+		}
+		headers = append(headers, "authorization: Bearer "+t.AccessToken)
+	}
+
+	refCtx := metadata.NewOutgoingContext(ctx, grpcurl.MetadataFromHeaders(headers))
+	cli := grpcreflect.NewClientAuto(refCtx, c.conn)
+	defer cli.Reset()
+
+	descSource := grpcurl.DescriptorSourceFromServer(ctx, cli)
+	in := strings.NewReader(payload)
+	options := grpcurl.FormatOptions{IncludeTextSeparator: true}
+	format := "text"
+	if !textFormat {
+		format = "json"
+	}
+	rf, formatter, err := grpcurl.RequestParserAndFormatter(grpcurl.Format(format), descSource, in, options)
+	if err != nil {
+		return nil, fmt.Errorf("grpcurl.RequestParserAndFormatter failed: %w", err)
+	}
+
+	o := new(bytes.Buffer)
+	h := &grpcurl.DefaultEventHandler{Out: o, Formatter: formatter}
```

Comment:
nit: it's more common to
```suggestion
	var o bytes.Buffer
	h := &grpcurl.DefaultEventHandler{Out: &o, Formatter: formatter}
```